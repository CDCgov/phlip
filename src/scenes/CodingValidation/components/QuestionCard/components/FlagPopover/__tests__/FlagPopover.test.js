import React from 'react'
import { FlagPopover } from '../index'
import { shallow } from 'enzyme'

const props = {
  userFlag: { notes: '', type: 0 },
  questionFlags: [],
  onSaveFlag: jest.fn(),
  user: {},
  disableAll: false
}

const redFlag = {
  id: 1,
  type: 3,
  notes: 'stop coding this questione ewqeqw erewrewrwweweeerewr ',
  raisedBy: {
    userId: 5,
    firstName: 'Kristin',
    lastName: 'Muterspaw'
  },
  raisedAt: '2019-04-25T16:57:44.6752779'
}

describe('CodingValidation -- FlagPopover component', () => {
  test('should render correctly', () => {
    expect(shallow(<FlagPopover {...props} />)).toMatchSnapshot()
  })
  
  describe('if there is a red flag on this question', () => {
    test('should disable regular flag form', () => {
      const wrapper = shallow(<FlagPopover {...props} questionFlags={[redFlag]} />)
      expect(wrapper.find('form').find('SimpleInput').props().disabled).toEqual(true)
      expect(wrapper.find('form').find('FlexGrid').at(0).childAt(0).props().choices[0].disabled).toEqual(true)
      expect(wrapper.find('form').find('FlexGrid').at(0).childAt(0).props().choices[1].disabled).toEqual(true)
    })
    
    test('should disable the "save" button in the regular flag form', () => {
      const wrapper = shallow(<FlagPopover {...props} questionFlags={[redFlag]} />)
      expect(wrapper.find('form').find('WithTheme(Button)').at(1).props().disabled).toEqual(true)
    })
    
    test('should disable the edit button if the current user is the user who made the red flag', () => {
      const wrapper = shallow(<FlagPopover {...props} questionFlags={[redFlag]} user={{ id: 5 }} />)
      const tHeader = wrapper.find('Table').childAt(0).find('TableCell').at(2).childAt(0)
      expect(tHeader.text()).toEqual('Edit')
      const tBody = wrapper.find('Table').childAt(1).find('TableCell').at(2).find('IconButton')
      expect(tBody.length).toEqual(1)
    })
    
    test('should not show the edit button and table column if the red flag was not created by current user', () => {
      const wrapper = shallow(<FlagPopover {...props} questionFlags={[redFlag]} user={{ id: 4 }} />)
      const tHeader = wrapper.find('Table').childAt(0).find('TableCell').at(2)
      expect(tHeader.length).toEqual(0)
      const tBody = wrapper.find('Table').childAt(1).find('TableCell').at(2).find('IconButton')
      expect(tBody.length).toEqual(0)
    })
  })
  
  describe('if there is no red flag, but there is a user regular flag', () => {
    test('the regular form should not be disabled', () => {
      const wrapper = shallow(<FlagPopover {...props} userFlag={{ type: 1, notes: 'these are notes.' }} />)
      expect(wrapper.find('form').at(1).find('SimpleInput').props().disabled).toEqual(false)
      expect(wrapper.find('form').at(1).find('FlexGrid').at(0).childAt(0).props().choices[0].disabled).toEqual(false)
      expect(wrapper.find('form').at(1).find('FlexGrid').at(0).childAt(0).props().choices[1].disabled).toEqual(false)
    })
    
    test('should populate the form with the user flag from props', () => {
      const wrapper = shallow(<FlagPopover {...props} userFlag={{ type: 1, notes: 'these are notes.' }} />)
      expect(wrapper.find('form').at(1).find('SimpleInput').props().value).toEqual('these are notes.')
      expect(wrapper.find('form').at(1).find('FlexGrid').at(0).childAt(0).props().selected).toEqual(1)
    })
    
    test('the red flag form should not be disabled', () => {
      const wrapper = shallow(<FlagPopover {...props} userFlag={{ type: 1, notes: 'these are notes.' }} />)
      expect(wrapper.find('form').at(0).find('SimpleInput').length).toEqual(1)
    })
    
    test('should automatically be in edit mode for red flag form', () => {
      const wrapper = shallow(<FlagPopover {...props} userFlag={{ type: 1, notes: 'these are notes.' }} />)
      expect(wrapper.state().inEditMode).toEqual(true)
    })
  })
})
