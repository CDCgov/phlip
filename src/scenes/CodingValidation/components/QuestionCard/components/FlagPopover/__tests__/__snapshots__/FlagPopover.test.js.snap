// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodingValidation -- FlagPopover component should render correctly 1`] = `
<FlexGrid
  align="center"
  circular={false}
  container={true}
  flex={true}
  justify="stretch"
  padding={0}
  raised={false}
  style={
    Object {
      "height": 24,
      "width": "unset",
    }
  }
  type="row"
>
  <Popover
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
    target={
      Object {
        "icon": <Icon
          className=""
        />,
        "id": "stop-coding-question",
        "style": Object {
          "maxHeight": 500,
          "paddingLeft": 15,
          "paddingRight": 15,
        },
        "tooltip": "Stop coding this question",
      }
    }
    title="Stop Coding This Question"
  >
    <FlexGrid
      align="center"
      circular={false}
      container={true}
      flex={false}
      justify="stretch"
      padding="10px 0 0"
      raised={false}
      style={
        Object {
          "flexWrap": "nowrap",
          "maxHeight": 500,
          "minHeight": 200,
          "minWidth": 500,
          "width": 500,
        }
      }
      type="column"
    >
      <form
        onSubmit={[Function]}
        style={
          Object {
            "alignSelf": "stretch",
            "flex": 1,
            "minWidth": 500,
            "width": 500,
          }
        }
      >
        <FlexGrid
          align="stretch"
          circular={false}
          container={false}
          flex={true}
          justify="stretch"
          padding={16}
          raised={false}
          style={Object {}}
          type="column"
        >
          <SimpleInput
            error={false}
            helperText=""
            id="flag-notes"
            label="Notes"
            multiline={false}
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Enter Notes"
            required={true}
            shrinkLabel={true}
            type="text"
            value=""
          />
        </FlexGrid>
      </form>
      <FlexGrid
        align="stretch"
        circular={false}
        container={true}
        flex={false}
        justify="stretch"
        padding={16}
        raised={false}
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
        type="row"
      >
        <WithTheme(Button)
          color="accent"
          onClick={[Function]}
          raised={false}
          value="Cancel"
        />
        <WithTheme(Button)
          color="accent"
          onClick={[Function]}
          raised={false}
          type="submit"
          value="Save"
        />
      </FlexGrid>
    </FlexGrid>
  </Popover>
  <Popover
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
    target={
      Object {
        "color": "#757575",
        "icon": <Icon
          className=""
        />,
        "id": "flag-question",
        "tooltip": "Flag this question",
      }
    }
    title="Flags"
  >
    <form
      onSubmit={[Function]}
    >
      <FlexGrid
        align="stretch"
        circular={false}
        container={false}
        flex={true}
        justify="stretch"
        padding={16}
        raised={false}
        style={
          Object {
            "minWidth": 450,
          }
        }
        type="column"
      >
        <WithStyles(RadioGroup)
          choices={
            Array [
              Object {
                "color": "#2E7D32",
                "disabled": false,
                "label": "Flag for analysis",
                "text": <FlexGrid
                  align="center"
                  circular={false}
                  container={true}
                  flex={false}
                  justify="stretch"
                  padding={0}
                  raised={false}
                  style={Object {}}
                  type="row"
                >
                  <WithTheme(Icon)
                    color="#2E7D32"
                    style={
                      Object {
                        "paddingRight": 5,
                      }
                    }
                  >
                    flag
                  </WithTheme(Icon)>
                  <span
                    style={
                      Object {
                        "color": "black",
                      }
                    }
                  >
                    Flag for analysis
                  </span>
                </FlexGrid>,
                "type": 1,
              },
              Object {
                "color": "#CE4A00",
                "disabled": false,
                "label": "Notify Coordinator",
                "text": <FlexGrid
                  align="center"
                  circular={false}
                  container={true}
                  flex={false}
                  justify="stretch"
                  padding={0}
                  raised={false}
                  style={Object {}}
                  type="row"
                >
                  <WithTheme(Icon)
                    color="#CE4A00"
                    style={
                      Object {
                        "paddingRight": 5,
                      }
                    }
                  >
                    flag
                  </WithTheme(Icon)>
                  <span
                    style={
                      Object {
                        "color": "black",
                      }
                    }
                  >
                    Notify coordinator
                  </span>
                </FlexGrid>,
                "type": 2,
              },
            ]
          }
          error={false}
          helperText=""
          label="Flag Type"
          onChange={[Function]}
          required={true}
          selected={0}
        />
      </FlexGrid>
      <FlexGrid
        align="stretch"
        circular={false}
        container={false}
        flex={true}
        justify="stretch"
        padding={16}
        raised={false}
        style={Object {}}
        type="column"
      >
        <SimpleInput
          disabled={false}
          error={false}
          helperText=""
          id="flag-notes"
          label="Notes"
          multiline={false}
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Enter Notes"
          required={true}
          shrinkLabel={true}
          type="text"
          value=""
        />
      </FlexGrid>
      <FlexGrid
        align="stretch"
        circular={false}
        container={true}
        flex={true}
        justify="flex-end"
        padding={16}
        raised={false}
        style={Object {}}
        type="row"
      >
        <WithTheme(Button)
          color="accent"
          onClick={[Function]}
          raised={false}
          type="button"
          value="Cancel"
        />
        <WithTheme(Button)
          color="accent"
          disabled={false}
          onClick={[Function]}
          raised={false}
          type="submit"
          value="Save"
        />
      </FlexGrid>
    </form>
  </Popover>
</FlexGrid>
`;
